cmd/thanos/receive.go

```diff
@@ -210,20 +206,43 @@ func runReceive(
 		})
 	}
 
-	level.Debug(logger).Log("msg", "setting up receive http handler")
+	level.Debug(logger).Log("msg", "setting up http handler")
 	{
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(
-			func() error {
-				if err := webHandler.Run(ctx); err != nil {
-					return fmt.Errorf("error starting web server: %s", err)
-				}
-				return nil
-			},
-			func(err error) {
-				cancel()
-			},
-		)
+
+		mux := http.NewServeMux()
+		registerMetrics(mux, reg)
+		registerProfile(mux)
+		readinessProber.RegisterInMux(mux)
+		webHandler.HandleInMux("/", mux)
+
+		listener, err := net.Listen("tcp", httpBindAddr)
+		if err != nil {
+			return errors.Wrapf(err, "listen HTTP on address %s", httpBindAddr)
+		}
+
+		// Monitor incoming connections with conntrack.
+		listener = conntrack.NewListener(listener,
+			conntrack.TrackWithName("http"),
+			conntrack.TrackWithTracing())
+
+		operationName := nethttp.OperationNameFunc(func(r *http.Request) string {
+			return fmt.Sprintf("%s %s", r.Method, r.URL.Path)
+		})
+
+		errlog := stdlog.New(log.NewStdlibAdapter(level.Error(logger)), "", 0)
```

Comment:
Could you move this into struct creation?

```
httpSrv := &http.Server{
			Handler:  nethttp.Middleware(opentracing.GlobalTracer(), mux, operationName),
			ErrorLog: stdlog.New(log.NewStdlibAdapter(level.Error(logger)), "", 0),
		}
```

Reply:
yep